src/OrleansRuntime/Core/GrainMethodInvoker.cs

```diff
@@ -0,0 +1,183 @@
+                var extensionMap = this.rootInvoker as IGrainExtensionMap;
+                IGrainExtension extension;
+                if (extensionMap != null && extensionMap.TryGetExtension(request.InterfaceId, out extension))
+                {
+                    implementationType = extension.GetType();
+                }
+                else
+                {
+                    implementationType = this.Grain.GetType();
+                }
+
+                // Get or create the implementation map for this object.
+                var implementationMap = interfaceToImplementationMapping.GetOrCreate(
+                    implementationType,
+                    request.InterfaceId);
+
+                // Get the method info for the method being invoked.
+                MethodInfo methodInfo;
+                implementationMap.TryGetValue(request.MethodId, out methodInfo);
+                return methodInfo;
+            }
+        }
+
+        /// <inheritdoc />
+        public object[] Arguments => request.Arguments;
+        
+        /// <inheritdoc />
+        public object Result { get; set; }
+
+        /// <inheritdoc />
+        public async Task Invoke()
+        {
+            // Execute each stage in the pipeline. Each successive call to this method will invoke the next stage.
+            // Stages which are not implemented (eg, because the user has not specified an interceptor) are skipped.
+            var numInterceptors = interceptors.Count;
+            if (stage < numInterceptors)
+            {
+                // Call each of the specified interceptors.
+                var interceptor = this.interceptors[stage];
+                stage++;
+                await interceptor.Invoke(this);
+                return;
+            }
+
+            if (stage == numInterceptors)
+            {
+                stage++;
+
+                // Grain-level invoker, if present.
+                var intercepted = this.Grain as IMethodInvocationInterceptor;
+                if (intercepted != null)
+                {
+                    await intercepted.Invoke(this);
+                    return;
+                }
+            }
+
+#pragma warning disable 618
+            if (stage == numInterceptors + 1)
+            {
+                stage++;
+
+                // Deprecated grain-level invoker, if present.
+                var intercepted = this.Grain as IGrainInvokeInterceptor;
+                if (intercepted != null)
+                {
+                    this.Result = await intercepted.Invoke(this.Method, this.request, this);
+                    return;
+                }
+            }
+#pragma warning restore 618
+
+            if (stage == numInterceptors + 2)
+            {
+                stage++;
+
+                // Deprecated silo-level invoker, if present.
+                var grain = this.Grain as IGrain;
+                if (grain != null && deprecatedInvokeInterceptor != null)
```

Comment:
Is the order correct? I would have imagined that the deprecated silo level interceptor came before the grain level interceptor